content/stellar-contracts/utils/math/wad.mdx

```diff
@@ -0,0 +1,404 @@
+---
+title: WAD Fixed-Point Math
+description: High-precision decimal arithmetic
+---
+
+# Overview
+
+The WAD library provides fixed-point decimal arithmetic for Soroban smart contracts with 18 decimal places of precision.
+It's designed specifically for DeFi applications where precise decimal calculations are critical,
+such as interest rates, exchange rates, and token pricing.
+
+It is a fixed-point representation where:
+- **1.0 is represented as `1_000_000_000_000_000_000` (10^18)**
+- **0.5 is represented as `500_000_000_000_000_000`**
+- **123.456 is represented as `123_456_000_000_000_000_000`**
+
+This allows precise decimal arithmetic using only integer operations, avoiding the pitfalls
+of floating-point arithmetic in smart contracts.
+
+# Why WAD?
+
+## Problems with Alternatives
+
+**Native Integers (`i128`, `u64`):**
+- No decimal support - `1/2 = 0` instead of `0.5`
+- Loss of precision in financial calculations
+- Requires manual scaling for each operation
+
+**Floating-Point (`f64`, `f32`):**
+- Non-deterministic behavior across platforms
+- Rounding errors that compound in financial calculations
+- Security vulnerabilities from precision loss
+
+## Why WAD is Better
+
+- **High Precision**: 18 decimals is more than sufficient for financial calculations
+- **Deterministic**: Same inputs always produce same outputs
+- **Gas Efficient**: Uses native `i128` arithmetic under the hood
+- **Battle-Tested**: Used in production by MakerDAO, Uniswap, Aave, and others
+- **Ergonomic**: Operator overloading makes code readable: `a + b * c`
+- **Type Safe**: NewType pattern prevents mixing scaled and unscaled values
+
+# Design Decisions
+
+## 1. Why 18 Decimals?
+
+**18 decimals was chosen for several reasons:**
+
+- **Battle-Tested Standard**: Most DeFi protocols on use 18 decimals
+- **Sufficient Precision**: Provides precision down to 10^-18, far exceeding practical financial needs
+- **Interoperability**: Makes it easier to port DeFi protocols from other ecosystems to Stellar
+- **Performance**: Fits in `i128` without overflow concerns for typical values
+
+**Example Precision:**
+```rust
+// Interest rate: 5.5% = 0.055
+let rate = Wad::from_ratio(&e, 55, 1000); // 55/1000 = 0.055
+
+// Time fraction: 1 day / 365 days ≈ 0.00274
+let time = Wad::from_ratio(&e, 1, 365); // 1/365 ≈ 0.00274
+
+// Interest = principal * rate * time
+let interest = principal * rate * time;
+// Precise to 18 decimal places!
+```
+
+## 2. NewType Pattern
+
+We use a NewType `struct Wad(i128)` instead of a type alias:
+
+```rust
+// ❌ BAD: Type alias
+type Wad = i128;
+
+// ✅ GOOD: NewType
+pub struct Wad(i128);
```

Comment:
emojis + GOOD/BAD is not the best choice imo, because it's not inherently bad to use a type alias, newtype just fits better here